ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java

```diff
@@ -706,6 +752,55 @@ protected void deleteBlob(final String blobIdStr, final String serviceId, Future
       routerCallback.onPollReady();
     }
 
+    /**
+     * Requests for a blob to be undeleted asynchronously and invokes the {@link Callback} when the request completes.
+     * @param blobIdStr The ID of the blob that needs to be undeleted in string form
+     * @param serviceId The service ID of the service undeleting the blob. This can be null if unknown.
+     * @param futureResult A future that would contain information about whether the undelete succeeded or not,
+     *                     eventually.
+     * @param callback The {@link Callback} which will be invoked on the completion of a request.
+     */
+    protected void undeleteBlob(final String blobIdStr, final String serviceId, FutureResult<Void> futureResult,
+        final Callback<Void> callback) {
+      // Can skip GET if we can determine this is not a metadata bllob
+      if (isMaybeMetadataBlob(blobIdStr)) {
+        Callback<GetBlobResultInternal> internalCallback = (GetBlobResultInternal result, Exception exception) -> {
+          if (exception != null) {
+            completeOperation(futureResult, callback, null, exception, false);
+          } else if (result.getBlobResult != null) {
+            exception = new RouterException(
+                "GET blob call returned the blob instead of just the store keys (before undelete)",
+                RouterErrorCode.UnexpectedInternalError);
+            completeOperation(futureResult, callback, null, exception, false);
+          } else {
+            List<String> blobIdStrs = new ArrayList<>();
+            blobIdStrs.add(blobIdStr);
+            if (result.storeKeys != null) {
+              result.storeKeys.forEach(key -> blobIdStrs.add(key.getID()));
+            }
+            // UndeleteManager undeletes all chunks in parallel
+            currentOperationsCount.addAndGet(blobIdStrs.size());
+            doUndeleteOperation(blobIdStrs, serviceId, futureResult, callback);
+          }
+        };
+
+        GetBlobOptions options = new GetBlobOptionsBuilder().operationType(GetBlobOptions.OperationType.All)
+            .getOption(GetOption.None)
```

Comment:
I know the logic is mostly from TtlUpdate but let's be careful on Undelete own logic here. Note that `GetOption.None` returns all blobs that are not expired and not deleted.  I wonder if we should use `Include_Deleted_Blobs` for undelete operation. Please double check `GetOption` enum.

Reply:
you are right, that was a mistake. updated.